### PR TITLE
add. new auxiliary module: behind_ingensec

### PR DESCRIPTION
**Pull request by mekhalleh**
This module can help you to discover the real IP address behind the BinarySec/IngenSec service.

## Verification Steps

  1. Install the module as usual
  2. Start msfconsole
  3. Do: `use auxiliary/gather/behind_ingensec`
  4. Do: `set hostname www.ingensec.com`
  5. Do: `run`

## Options

  **HOSTNAME**

  This is the hostname [fqdn] on which the website responds. But this can also be a domain.

    msf auxiliary(gather/behind_ingensec) > set hostname www.ingensec.com
    --or--
    msf auxiliary(gather/behind_ingensec) > set hostname ingensec.com

  **RPORT**

  The target TCP port on which the protected website responds. Default: 443

  **SSL**

  Negotiate SSL/TLS for outgoing connections. Default: true

  **URIPATH**

  The URI path on which to perform the page comparison. Default: '/'

  **Poxies**

  A proxy chain of format type:host:port[,type:host:port][...]. It's optional.

  **THREADS**

  Number of concurent threads needed for DNS enumeration. Default: 15

  **WORDLIST**

  Name list required for DNS enumeration. Default: ~/metasploit-framework/data/wordlists/namelist.txt

## Advanced options

  **VERBOSE**

  You can also enable the verbose mode to have more information displayed in the console.

### For auditing purpose

  If successful, you must be able to obtain the IP address of the website as follows:

  ```
  [*] Passive gathering information...
  [*]  * PrePost SEO: 1 IP address found(s).
  [*]  * DNS Enumeration: 12 IP address found(s).
  [*] 
  [*] Clean binarysec/ingensec server(s)...
  [+]  * TOTAL: 3 IP address found(s) after cleaning.
  [*] 
  [*] Bypass BinarySec/IngenSec is in progress...
  [*]  * Initial request to the original server for comparison
  [*]  * Trying: XXX.XXX.XXX.XXX
  [+] A direct-connect IP address was found: XXX.XXX.XXX.XXX
  [*] Auxiliary module execution completed
  ```

  For example:

  For some reason you may need to change the URI path to interoperate with other than the index page.
  To do this specific thing:

  ```
  msf > use auxiliary/gather/behind_ingensec
  msf auxiliary(gather/behind_ingensec) > set HOSTNAME www.ingensec.com
  hostname => www.ingensec.com
  msf auxiliary(gather/behind_ingensec) > set URIPATH /fct_securite.html
  uripath => /fct_securite.html
  msf auxiliary(gather/behind_ingensec) > run
  ...
  ```

01-demo
02-demo
03-demo

### CODE REVIEW

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    rescue ::Rex::ConnectionRefused, Rex::ConnectionError
+      return(false)
+    end
+    sock.close
+    return(true)
+  end
+
+  def do_grab_domain_ip_history(hostname, proxies)
+    begin
+      cli = Rex::Proto::Http::Client.new('www.prepostseo.com', 443, {}, true, nil, proxies)
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: Consider using `send_request_cgi` ?

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+  def do_check_tcp_port(ip, port, proxies)
+    begin
+      sock = Rex::Socket::Tcp.create(
+        'PeerHost' => ip,
+        'PeerPort' => port,
+        'Proxies'  => proxies
+      )
+    rescue ::Rex::ConnectionRefused, Rex::ConnectionError
+      return(false)
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: Just code `return false` as [ruby style guide](https://github.com/bbatsov/ruby-style-guide) suggest.

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    end
+    sock.close
+    return(true)
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: Just code `return true` to keep consistent.

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        'uri'    => '/domain-ip-history-checker',
+        'method' => 'POST',
+        'data'   => "url=#{hostname}&submit=Check+Reverse+Ip+Domains"
+      })
+      response = cli.send_recv(request)
+      cli.close
+
+    rescue ::Rex::ConnectionError, Errno::ECONNREFUSED, Errno::ETIMEDOUT
+      print_error('HTTP connection failed to PrePost SEO website.')
+      return(false)
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: Also, `return false`

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    end
+
+    html  = response.get_html_document
+    table = html.css('table.table').first
+    rows  = table.css('tr')
+
+    arIps = Array.new
+    rows.each_with_index.map do | row, index |
+      row = /(\d*\.\d*\.\d*\.\d*)/.match(row.css('td').map(&:text).to_s)
+      unless row.nil? then
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: No need to add `then` in the end,

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        arIps.push(row)
+      end
+    end
+
+    if arIps.empty? then
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: No need to add `then` in the end,

File: modules/auxiliary/gather/behind_ingensec.rb [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        end
+        t.map(&:join)
+
+      rescue ::Timeout::Error
+      ensure
+        t.each { |x| x.kill rescue nil }
+      end
+    end
+
+    if arIps.empty? then
```
Comment (Green-m) [inline modules/auxiliary/gather/behind_ingensec.rb]: No need to add `then` in the end,

**Comment by Green-m** [comment]
A nice module!

But before we merge it, for the code style consistent, some little problems should be fixed.

1. Do not use then for multi-line if/unless.

```
# bad
if some_condition then
  # body omitted
end

# good
if some_condition
  # body omitted
end
```

2. Can omit parentheses for methods that have "keyword" status in Ruby, but are not declarative:

```
puts temperance.age
system 'ls'
return false
```

3. Use `snake_case` for symbols, methods, and variables.

```
# bad
arIps = Array.new

# good 
ar_ips = []

```

For more:
https://github.com/rubocop-hq/ruby-style-guide 
https://github.com/rapid7/metasploit-framework/blob/master/CONTRIBUTING.md

File: modules/auxiliary/gather/behind_ingensec.rb (hunk 1 above)
Comment (mekhalleh) [reply modules/auxiliary/gather/behind_ingensec.rb]: thanks for the review.
I'm not sure I understand that point. Can you tell me more about this?

Comment (Green-m) [reply modules/auxiliary/gather/behind_ingensec.rb]: `send_request_cgi` and `send_request_raw` are the good wrapper for CGI and arbitrary HTTP requests, which could make the code cleaner and more readable.

See doc [here](https://rapid7.github.io/metasploit-framework/api/Msf/Exploit/Remote/HttpClient.html#send_request_cgi-instance_method)

Comment (mekhalleh) [reply modules/auxiliary/gather/behind_ingensec.rb]: I keep this reference for later (I have to implement other features in the future).

I made a code update and a commit.

PS: add. new auxiliary module: behind_cloudflare # 10481 is also based on bypass, if you can also do a review? I anticipated the synthaxic corrections.

Thanks.

File: behind_ingensec.rb [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    end
+    sock.close
+    return true
+  end
+
+  def do_grab_domain_ip_history(hostname, proxies)
+    begin
+      cli = Rex::Proto::Http::Client.new('www.prepostseo.com', 443, {}, true, nil, proxies)
+      cli.connect
+
```
Comment (sempervictus) [inline behind_ingensec.rb]: Good use of the namespace you have, but as a language construct we use mixins for this. _Exploit::Remote::Http_, I believe is what would be included to abstract this part. Take look at some other http scanners to see how the convenience methods handle this and underlying mechanisms like socket registration.

File: behind_ingensec.rb [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    arIps = []
+    rows.each_with_index.map do | row, index |
+      row = /(\d*\.\d*\.\d*\.\d*)/.match(row.css('td').map(&:text).to_s)
+      unless row.nil?
+        arIps.push(row)
+      end
+    end
+
+    if arIps.empty?
```
Comment (sempervictus) [inline behind_ingensec.rb]: Please no camel-case. There are enough languages which hurt the eyes out there. As mentioned above, convention is snake-case.

File: behind_ingensec.rb [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    return arIps
+  end
+
+  ## auxiliary/gather/enum_dns.rb
+  def do_dns_enumeration(domain, threads)
+    wordlist = datastore['WORDLIST']
+    return if wordlist.blank?
+    threads  = 1 if threads <= 0
+
+    queue    = []
```
Comment (sempervictus) [inline behind_ingensec.rb]: Queue.new?

File: behind_ingensec.rb [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+      print_error("Query #{domain} DNS #{type} - exception: #{e}")
+      return nil
+    end
+  end
+
+  ## TODO: Improve on-demand response (send_recv? Timeout).
+  def do_simple_get_request_raw(host, port, ssl, host_header=nil, uri, proxies)
+    begin
+      http    = Rex::Proto::Http::Client.new(host, port, {}, ssl, nil, proxies)
+      http.connect
```
Comment (sempervictus) [inline behind_ingensec.rb]: Again good use of primitives but this is implemented with nuanced considerations for the mess that is http in the exploit::http namespace

File: behind_ingensec.rb [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+        vprint_status(" * Trying: #{ip}")
+
+        ret_val = do_check_bypass(
+            head,
+            datastore['HOSTNAME'],
+            ip,
+            datastore['URIPATH'],
+            datastore['PROXIES']
+        )
+        break if ret_val.eql? true
```
Comment (sempervictus) [inline behind_ingensec.rb]: Record notes or vulns maybe instead of console-only?

**Review by sempervictus — CHANGES_REQUESTED** [review]
Other than the general notion of "try to use what the framework brings because it will save you and your users pain," its readable and looks sane. Will try it out when a relevant use-case comes up. Thanks

File: behind_ingensec.rb (hunk 10 above)
Comment (mekhalleh) [reply behind_ingensec.rb]: This is implemented into auxiliary/gather/enum_dns.rb I do not have enough feedback in this language to judge.

File: behind_ingensec.rb (hunk 12 above)
Comment (mekhalleh) [reply behind_ingensec.rb]: it's planned :)

File: behind_ingensec.rb (hunk 8 above)
Comment (mekhalleh) [reply behind_ingensec.rb]: Thank you, I have already made this remark. in fact it's my second code Ruby and I must admit that it's a bit confusing.

I did not find a use case in this context w/ scanner then I relied on modules/auxiliary/gather/censys_search.rb

File: behind_ingensec.rb (hunk 11 above)
Comment (mekhalleh) [reply behind_ingensec.rb]: Thank you, I have already made this remark. in fact it's my second code Ruby and I must admit that it's a bit confusing.

I did not find a use case in this context w/ scanner then I relied on modules/auxiliary/gather/censys_search.rb

**Comment by mekhalleh** [comment]
add censys as data source
rewrite the output like behind_cloudflare module
doc. update

02-demo

**Comment by bcoles** [comment]
Please see the review comments in #10481. Many of the comments also apply to this module.

**Comment by mekhalleh** [comment]
fixed.

**Review by bcoles — DISMISSED** [review]
It appears you have accidentally included the `behind_ingensec.rb` file twice

**Comment by mekhalleh** [comment]
@msjenkins-r7 recheck ;)

**Comment by mekhalleh** [comment]
**CLOSE BECAUSE THE NEW MODULE https://github.com/rapid7/metasploit-framework/pull/12234 REPLACES THIS**

**Pull request closed without merging**